File: src/cnft/leaf-event.ts
// LeafSchemaEvent decoder — the authoritative new-state announcement
// Bubblegum emits as an inner CPI to the noop program on every
// leaf-mutating instruction. Parsing this lets us track verifyCreator
// / verifyCollection / updateMetadata and similar ixs whose new state
// can't be reconstructed from the outer ix args alone.
//
// Wire layout (borsh-packed, per mpl-bubblegum's LeafSchemaEvent):
//
//     [0]    u8 event_type          — BubblegumEventType (1 = LeafSchemaEvent)
//     [1]    u8 version             — historical; ignored on our side
//     [2..]  LeafSchema             — V1 (0x00 disc) or V2 (0x01 disc)
//     [..]   InstructionName        — u8 enum identifying the emitting ix
//
// The same noop program (SPL Noop) is also the sink for
// spl-account-compression's ChangeLogEvent, which has a different
// layout — we distinguish by the first byte (event_type==1) cheaply
// before attempting a full decode.

import {
  getAddressEncoder,
  type Address,
} from "@solana/kit";
import { getBubblegumEventTypeDecoder } from "../generated/bubblegum/types/bubblegumEventType.js";
import { getInstructionNameDecoder } from "../generated/bubblegum/types/instructionName.js";
import {
  getLeafSchemaDecoder,
  getLeafSchemaEncoder,
} from "../generated/bubblegum/types/leafSchema.js";

// Program IDs for the two noops Bubblegum has emitted events through.
// SPL Noop is the historical + current default; MPL Noop is used by the
// V2 ix family introduced in recent Bubblegum versions. We accept both.
export const SPL_NOOP_PROGRAM_ID =
  "noopb9bkMVfRPU8AsbpTUg8AQkHtKwMYZiFUjNRtMmV" as Address;
export const MPL_NOOP_PROGRAM_ID =
  "mnoopTCrg4p8ry25e4bcWA9XZjbNjMTfgYVGGEdRsf3" as Address;
export const NOOP_PROGRAM_IDS: ReadonlySet<string> = new Set([
  SPL_NOOP_PROGRAM_ID as unknown as string,
  MPL_NOOP_PROGRAM_ID as unknown as string,
]);

/**
 * What we hand back on a successful decode. All 32-byte values are
 * passed through as Uint8Array so downstream logic doesn't have to
 * re-decode base58 to compare bytes.
 */
export interface LeafSchemaEventDecoded {
  /** Which on-chain instruction produced this event. Informational — applied code doesn't branch on it. */
  op: string;
  /** The authoritative new leaf schema. */
  schema:
    | {
        kind: "V1";
        id: Uint8Array;
        owner: Uint8Array;
        delegate: Uint8Array;
        nonce: bigint;
        dataHash: Uint8Array;
        creatorHash: Uint8Array;
      }
    | {
        kind: "V2";
        id: Uint8Array;
        owner: Uint8Array;
        delegate: Uint8Array;
        nonce: bigint;
        dataHash: Uint8Array;
        creatorHash: Uint8Array;
      };
}

const addrEncoder = getAddressEncoder();

function addressToBytes(addr: Address): Uint8Array {
  return addrEncoder.encode(addr) as Uint8Array;
}

/**
 * Decode LeafSchemaEvent bytes. Returns null for:
 *   - Non-LeafSchemaEvent noop payloads (wrong first byte)
 *   - Truncated / malformed bytes
 *
 * Never throws — callers treat null as "this inner ix isn't our event,
 * skip it." This is hot: every tx has multiple noop CPIs and we
 * attempt-decode each one.
 */
export function decodeLeafSchemaEvent(
  data: Uint8Array,
): LeafSchemaEventDecoded | null {
  // Cheap discriminator check before spending any decoder cycles. The
  // first byte is BubblegumEventType — 1 = LeafSchemaEvent, 0 =
  // Uninitialized. ChangeLogEvent from spl-account-compression has its
  // own layout that won't start with a `1` in the event-type slot.
  if (data.length < 3) return null;
  if (data[0] !== 1) return null;

  try {
    const eventType = getBubblegumEventTypeDecoder().decode(data.subarray(0, 1));
    if (eventType !== 1) return null;

    // Byte 1 is the historical `version` — read and discard.
    let offset = 2;

    const leafDec = getLeafSchemaDecoder();
    const schema = leafDec.decode(data.subarray(offset));

    // LeafSchema is a discriminated union, so its encoded size varies
    // between V1 and V2. Re-encode the decoded value to find how many
    // bytes it consumed, then advance past it to read the trailing
    // InstructionName. Kit doesn't expose "bytes read" on a decoder
    // directly, so this round-trip is the simplest correct approach.
    const consumed = getLeafSchemaEncoder().encode(schema).length;
    offset += consumed;

    if (offset >= data.length) return null;
    const opRaw = getInstructionNameDecoder().decode(
      data.subarray(offset, offset + 1),
    );
    const opName = typeof opRaw === "number" ? `Unknown(${opRaw})` : String(opRaw);

    return {
      op: opName,
      schema:
        schema.__kind === "V1"
          ? {
              kind: "V1",
              id: addressToBytes(schema.id),
              owner: addressToBytes(schema.owner),
              delegate: addressToBytes(schema.delegate),
              nonce: schema.nonce,
              dataHash: new Uint8Array(schema.dataHash),
              creatorHash: new Uint8Array(schema.creatorHash),
            }
          : {
              kind: "V2",
              id: addressToBytes(schema.id),
              owner: addressToBytes(schema.owner),
              delegate: addressToBytes(schema.delegate),
              nonce: schema.nonce,
              dataHash: new Uint8Array(schema.dataHash),
              creatorHash: new Uint8Array(schema.creatorHash),
            },
    };
  } catch {
    return null;
  }
}

File: crates/service/src/cnft/leaf_event.rs
//! LeafSchemaEvent decoder — the authoritative new-state payload
//! Bubblegum emits via inner CPI to the noop program on every
//! leaf-mutating instruction.
//!
//! Wire layout is mpl-bubblegum's `LeafSchemaEvent` struct,
//! Borsh-serialized:
//!
//! ```text
//! event_type  : BubblegumEventType (Borsh enum → u8)
//! version     : Version             (Borsh enum → u8)
//! schema      : LeafSchema          (Borsh enum, V1 or V2)
//! leaf_hash   : [u8; 32]            (final new leaf hash)
//! ```
//!
//! We deserialize the whole struct via mpl-bubblegum's own type
//! definitions — anchor-generated + `BorshDeserialize`-derived, so we
//! inherit any layout changes they ship without hand-maintaining
//! parallel logic. The first-byte fast-path lets us cheaply reject
//! non-LeafSchemaEvent noop payloads (e.g. spl-account-compression
//! ChangeLogEvents) before spending decoder cycles.

use borsh::BorshDeserialize;
use mpl_bubblegum::types::{BubblegumEventType, LeafSchema};
use mpl_bubblegum::LeafSchemaEvent;

/// Two noop program IDs in the wild. SPL Noop has been Bubblegum's
/// historical sink; MPL Noop ships with the V2 ix family.
pub const SPL_NOOP_PROGRAM_ID: &str = "noopb9bkMVfRPU8AsbpTUg8AQkHtKwMYZiFUjNRtMmV";
pub const MPL_NOOP_PROGRAM_ID: &str = "mnoopTCrg4p8ry25e4bcWA9XZjbNjMTfgYVGGEdRsf3";

/// True when the given program id is one of the two noop sinks.
#[must_use]
pub fn is_noop_program(program_id: &str) -> bool {
    program_id == SPL_NOOP_PROGRAM_ID || program_id == MPL_NOOP_PROGRAM_ID
}

/// Decoded LeafSchemaEvent in our service-layer shape. All Pubkey
/// fields are unpacked to `[u8; 32]` so downstream logic compares
/// against stored state without re-encoding through base58.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct LeafSchemaEventDecoded {
    pub schema: DecodedLeafSchema,
    /// The final leaf hash emitted by Bubblegum — useful to
    /// cross-check our recomputed leaf_hash against on-chain truth.
    pub leaf_hash: [u8; 32],
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum DecodedLeafSchema {
    V1 {
        id: [u8; 32],
        owner: [u8; 32],
        delegate: [u8; 32],
        nonce: u64,
        data_hash: [u8; 32],
        creator_hash: [u8; 32],
    },
    /// V2 variants carry extra fields (collection_hash, asset_data_hash,
    /// flags). We don't track them yet; represented distinctly so the
    /// parser surfaces `Unsupported` rather than silently dropping.
    V2 {
        id: [u8; 32],
        owner: [u8; 32],
        delegate: [u8; 32],
        nonce: u64,
        data_hash: [u8; 32],
        creator_hash: [u8; 32],
    },
}

/// Decode noop CPI data as a LeafSchemaEvent. Returns `None` for:
/// - Non-LeafSchemaEvent noop payloads (wrong event-type byte)
/// - Truncated or malformed bytes
/// - Anything that doesn't Borsh-parse cleanly
///
/// Never panics — this is hot, called once per inner noop CPI per tx.
#[must_use]
pub fn decode_leaf_schema_event(data: &[u8]) -> Option<LeafSchemaEventDecoded> {
    // Cheap discriminator check before decoding. BubblegumEventType's
    // Borsh wire form for LeafSchemaEvent is 0x01; ChangeLogEvent from
    // spl-account-compression has its own layout we want to skip.
    if data.is_empty() || data[0] != BubblegumEventType::LeafSchemaEvent as u8 {
        return None;
    }

    let event = LeafSchemaEvent::try_from_slice(data).ok()?;
    let schema = match event.schema {
        LeafSchema::V1 {
            id,
            owner,
            delegate,
            nonce,
            data_hash,
            creator_hash,
        } => DecodedLeafSchema::V1 {
            id: id.to_bytes(),
            owner: owner.to_bytes(),
            delegate: delegate.to_bytes(),
            nonce,
            data_hash,
            creator_hash,
        },
        LeafSchema::V2 {
            id,
            owner,
            delegate,
            nonce,
            data_hash,
            creator_hash,
            ..
        } => DecodedLeafSchema::V2 {
            id: id.to_bytes(),
            owner: owner.to_bytes(),
            delegate: delegate.to_bytes(),
            nonce,
            data_hash,
            creator_hash,
        },
    };

    Some(LeafSchemaEventDecoded {
        schema,
        leaf_hash: event.leaf_hash,
    })
}

impl LeafSchemaEventDecoded {
    /// Convert into the service-layer `NoopOverride` shape when the
    /// schema is V1. V2 returns None — the parser turns that into a
    /// `ParseError::Unsupported`.
    #[must_use]
    pub fn as_v1_override(&self) -> Option<crate::cnft::types::NoopOverride> {
        match &self.schema {
            DecodedLeafSchema::V1 {
                nonce,
                owner,
                delegate,
                data_hash,
                creator_hash,
                ..
            } => Some(crate::cnft::types::NoopOverride {
                leaf_index: *nonce,
                nonce: *nonce,
                owner: *owner,
                delegate: *delegate,
                data_hash: *data_hash,
                creator_hash: *creator_hash,
            }),
            DecodedLeafSchema::V2 { .. } => None,
        }
    }
}
